airbyte-cdk/python/airbyte_cdk/sources/source.py

```diff
@@ -66,13 +67,23 @@ def read_state(self, state_path: str) -> List[AirbyteStateMessage]:
                     parsed_state_messages.append(parsed_message)
                 return parsed_state_messages
             else:
-                # Existing connectors that override read() might not be able to interpret the new state format. We temporarily
-                # send state in the old format for these connectors, but once all have been upgraded, this block can be removed
-                # vars(self.__class__) checks if the current class directly overrides the read() function
-                if "read" in vars(self.__class__):
-                    return state_obj
+                return self._emit_legacy_state_format(state_obj)
+        return self._emit_legacy_state_format({})
+
+    def _emit_legacy_state_format(self, state_obj) -> Union[List[AirbyteStateMessage], MutableMapping[str, Any]]:
+        """
+        Existing connectors that override read() might not be able to interpret the new state format. We temporarily
+        send state in the old format for these connectors, but once all have been upgraded, this method can be removed,
+        and we can then emit state in the list format.
+        """
+        # vars(self.__class__) checks if the current class directly overrides the read() function
+        if "read" in vars(self.__class__):
+            return defaultdict(dict, state_obj)
```

Comment:
Could you share why you call `defaultdict` with a `dict` default factory?

Reply:
Great question! Not sure if it actually makes a difference, but the original code before the changes also wrapped it in `defaultdict` so I wanted to preserve the same functionality. Reference to prior changes: https://github.com/airbytehq/airbyte/blob/055763ec73e2dd260715b8b108664cf96dae5d23/airbyte-cdk/python/airbyte_cdk/sources/source.py#L49